aips/aip-55.md

```diff
@@ -0,0 +1,97 @@
+
+This AIP introduces a new `TransactionAuthenticator` called `SingleSender` that supports two `AccountAuthenticator`s, SingleKeyAuthenticator and MultiKeyAuthenticator, which supports a single key and a k-of-n multi-key, respectively. These authenticators decouple the proof type from both the `TransactionAuthenticator` and `AccountAuthenticator` simplifying the addition of new cryptographic proofs for account authentication.
+
+### Goals
+
+* Eliminate duplication of code across `TransactionAuthenticator`, `AccountAuthenticator`, and `AuthencticationKey` by providing a single means to represent the identity therein.
+* Make adding new cryptographic protocols consistent across multi-sender and single-sender transactions. Current single-sender transactions expect a specific type of cryptographic algorithm in the `TransactionAuthenticator` where as multi-sender transactions expect them in the `AccountAuthenticator`.
+* Deliver a unified framework for MultiKey authentication that works across both single and multi-sender transactions.
+
+## Alternative solutions
+
+This work could have focused purely on multi-key solutions that work across the board. This would have eliminated the optimization for single-sender applications. Doing so increases the burden for creating adding cryptographic algorithms for authorization.
+
+Instead of doing this at an account level, we could have used Multisigv2. In Multisigv2, multiple on-chain accounts share a single account. This is similar to the model here, where multiple keys share the same account. The caveat being that if keys are misplaced in a Multisigv2, the assets within those accounts are lost. Those accounts require minimally gas fees, so there would be inevitable loss.
+
+## Specification
+
+### Data Structures
+
+```
+TransactionAuthenticator::SingleSender { sender: AccountAuthenticator }
+AccountAuthenticator::SingleKey { authenticator: SingleKeyAuthenticator }
+AccountAuthenticator::MultiKey { authenticator: MultiKeyAuthenticator }
+
+pub struct SingleKeyAuthenticator {
+    public_key: AnyKey,
+    signature: AnySignature,
+}
+
+pub struct MultiKeyAuthenticator {
+    public_keys: MultiKey,
+    signatures: Vec<AnySignature>,
+    signatures_bitmap: aptos_bitvec::BitVec,
+}
+
+pub enum AnySignature {
+    Ed25519 {
+        signature: Ed25519Signature,
+    },
+    Secp256k1Ecdsa {
+        signature: secp256k1_ecdsa::Signature,
+    },
+}
+
+pub enum AnyKey {
+    Ed25519 {
+        public_key: Ed25519PublicKey,
+    },
+    Secp256k1Ecdsa {
+        public_key: secp256k1_ecdsa::PublicKey,
+    },
+}
+```
+
+New cryptographic algorithms can be added by adding an appropriate data structure and verify method under `AnyKey` and `AnySignature`.
+
+The `MultiKeyAuthenticator` supports up to 255 keys and up to 32 signatures, as 32 signatures are the maximum allowed in a single transaction.
```

Comment:
Can you clarify why is 32 the maximum? Is it because you assume signatures are, say 32 bytes, so $32 \times 32$ bytes is $960$ bytes which gets close to a potential 1KiB limit in the code?

Reply:
Also, does this mean in a $t$-out-of-$n$ multikey address cannot have $t > 32$?